src/nunavut/lang/cpp/support/variable_length_array.hpp

```diff
@@ -39,47 +70,274 @@ class VariableLengthArray
         , alloc_()
     {}
 
-    explicit VariableLengthArray(const Allocator& alloc) noexcept(std::is_nothrow_copy_constructible<Allocator>::value)
+    VariableLengthArray(std::initializer_list<T> l) noexcept
```

Comment:
The design of the built-in VLA walks a fine line between utility and avoiding over-generalization. That is, we're not Boost; we're not trying to provide a great VLA type you should use everywhere. Instead, I'm trying to provide just enough support that the default c++ types are usable and that are optimized for our serialization logic. For example, if you used this type in generic code you might get pretty frustrated by `push_back` not resizing the array when you are up against the current capacity. There'd be a lot of `if at capacity, grow, then push_back` logic polluting your beautiful code. For Nunavut, you are only consuming the output of our deserializers so you wouldn't be exposed to this wart. Internally, this allows us to remove a lot of logic from push_back since our deserializers tend to optimize themselves in ways that allow a single `reserve` call.

This design philosophy means we don't provide a lot of support for going to and from STL containers. The copy and move constructors, for example, do not resolve to any container supporting the internal concepts. Instead we only allow copy, move, compare between `VariableLengthArray` object and only if they are exactly the same Type, MaxSize, and Allocator.

Reply:
Given that the capacity (the maximum size) is known at compile time and that we are moving towards a more static design, do we not want to call `reserve(MaxSize)` in the default constructor so that the user doesn't have to deal with this manually afterward? If this default behavior is undesirable we could provide a constructor overload that takes a tag type that suppresses this behavior or simply takes the desired size of the array like std::vector does.

Reply:
No. I considered this design but then why have a VLA type at all and not just use arrays with a size property. The end goal is to allow some sort of allocation to occur that means a constrained environment doesn't need to keep the max size of all variable length array types it deserializes allocated.

Reply:
Okay, but I am uneasy about sacrificing usability for performance. Can you please elaborate on why this is not a case of premature optimization? If we were to embed a call to `reserve()` from `push_back()` with some standard resize policy (e.g., golden ratio), what would be the actual downside aside from the marginal increase in complexity?

>For Nunavut, you are only consuming the output of our deserializers so you wouldn't be exposed to this wart.

You would be exposed to it during serialization and also if DSDL objects make their way into the business logic of the application.